### ISSUE TITLE
Feedback and questions

### ISSUE THREAD

**Issue by Lakedaemon** [aler9/rtsp-simple-server#162]
## Which version are you using?

v0.13.0

## Question

I would like to stream 2 cams totaling 100kB/s through a vpn tunnel (that goes through a 50Mb/s 4 g connection) to a vps server (with 250Mb/s) so that 45 students can watch both streams

I tried this configuration : 

paths:
 proxied1:
    source: rtsp://vpnClientIP1:554/ipc1
    sourceProtocol: tcp
    sourceOnDemand: yes
 proxied2:
    source: rtsp://vpnClientIP2:554/ipc2
    sourceProtocol: tcp
    sourceOnDemand: yes

But, when the 40+ students connected to the server, we had lots of lost paquets...
(the video was switching on and off)
I think that I am doing something really wrong with the configuration of rtsp-simple-server

Is this configuration opening 2 * 40 connections between the server and the cams ?
(I want to open only 2)

If yes, what should I do to only open 2 conections between the server and the cams ? not using proxy ? restream ? 

Is there something I could do to lessen the latency of this rig ?
(the cams are using vlc or ffmpeg, h264 or h265, rtsp or tmtp or other protocols)
I think that I can get to 2s of latency by using rtsp + h265. Can I do better ?

At one point, I will want to record the streams, as soon as the cam are turned on. 
How should I proceed ? 
(removing sourceOnDeman I guess ? is it still possible using rtsp ? or should I turn to rtmp)

Apparently, it rtsp-server manages to connect to the cam with tcp, if I put udp (or automatic) instead, it fails

Best regards,
Your software is simple to use and shows a lot of promises (I am trying to use it to convince math teachers to teach from afar, by using it)

**Comment by Lakedaemon** [comment]
It looks like I am doing the right thing though : my situation is exactly what is told on both points : 

- when there are multiple users that are receiving a stream and the bandwidth is limited; the proxy is used to receive the stream once. Users can then connect to the proxy instead of the original source.
- when there's a NAT / firewall between a stream and the users; the proxy is installed on the NAT and makes the stream available to the outside world.

and I'm using this syntax : 
paths:
  proxied1:
    source: rtsp://url1

  proxied2:
    source: rtsp://url1

I am steamed :/

**Comment by Lakedaemon** [comment]
Maybee this is because of the udp protocol (there are some lost packets), the codec h265 and the heavy compression I use to stream only 50ko/s for a 1080p view of a blackboard. As I don't have many key frames, If I lose a tiny packet for the key frame, it degrades the video replay for 2s.

I'll try, experiment and report back (as simple-rtsp-server looks like it is doing it's part correctly)

**Comment by aler9** [comment]
Hello,
At the moment, to maximize performance, readings and writings are done in sequence, not in parallel, in the following way (pseudocode):

```
while true {
    read a frame from publisher
    for each reader {
        write frame to reader
    }
}
```

This means that if even a single reader has a laggy connection, all the others are delayed, and that's the reason for the UDP protocol not working well with high bitrates.

The model that will be approached in the future will be

```
while true {
    read a frame from publisher
    for each reader {
        put frame in the reader's write queue
    }
}
```

for each reader, a write queue will be instantiated:

```
while true {
    read a frame from queue
    write frame to reader
}
```

The problem of this work is that i have to find the way to keep the same level of performance - at the moment, the new approach consumes 100% more CPU, so i can't deploy it.

**Comment by Lakedaemon** [comment]
This may be a completly stupid suggestion, because I'm coming from the kotlin world (not go) and I don't know the udp protocol.

What about having a coroutine for each reader and for the publisher
When the publisher doesn't have a frame ready, everybody sleeps (no cpu work)
When the publisher produces a frame, it sends it to the reader coroutines through a channel,
coroutines wake up and do their work and go back to sleep
Coroutines are launched on a thread pool, which does the parallel work for you

In the kotlin/Java world, we have Flow and RxJava (reactive flows) which kind of handle just that. Maybee you can take inspiration, there.

**Comment by Lakedaemon** [comment]
I had a very quick look at the code and I found this : 

// Write writes a UDP packet.
func (s *Server) Write(buf []byte, addr *net.UDPAddr) {
	s.writeMutex.Lock()
	defer s.writeMutex.Unlock()

	s.pc.SetWriteDeadline(time.Now().Add(s.writeTimeout))
	s.pc.WriteTo(buf, addr)
}

I am probably wrong (remmeber that I know nothing of go/udp) but

If there is a lock contention befere every frame write, isn't it going to kill performance ?

Is it not possible to use a read only/immutable frame/content that you can use/reuse in all threads ?
If you always work with the same reader in the same thread, there would be no need for a mutex

In kotlin, I had to use coroutines with a few threads and as, coroutine might jump threads (when they stop and resumes),
I implemented a thread->thread unsafe structure (read only/immutable) map that allowed me : 
1) At the start of the coroutine workload, get with a map call a thread unsafe structure associated with the current thread
2) (in the current thread), in the coroutine workload : use the thread unsafe structure (works just fine as it is only used in the right thread)

**Comment by aler9** [comment]
> I had a very quick look at the code and I found this

That mutex is inserted due to the fact that SetWriteDeadline() has a strange behavior when called by multiple routines in parallel, sometimes it even freezes the program - but it may be moved in a dedicated routine, and in this way writeMutex would not be necessary, maybe we can do some tests about it and increase performance further.

> What about having a coroutine for each reader and for the publisher
> When the publisher doesn't have a frame ready, everybody sleeps (no cpu work)
> When the publisher produces a frame, it sends it to the reader coroutines through a channel,
> coroutines wake up and do their work and go back to sleep

this was exactly the original solution, but then benchmarks showed that a mutex-based solution with a single, shared buffer is the most performant solution with the Go language, probably because the garbage collector don't need to run often (if there are many buffers, old ones must be garbage-collected). After a while we discovered problems like the one described in this issue, that happens when there are many readers connected with variable bandwidth, and like the UDP one. They happen because a single reader is slow and blocks every other readers and the publisher.

Something like this

```go
func publisherRoutine() {
    for {
        buf := make([]byte, 2048)
        readFrame(buf)

        for _, r := range reader {
            r.writeQueue <- buf
        }
    }
}

func readerRoutine() {
    for _, buf := range r.writeQueue {
        r.writeFrame(buf)
    }
}
```

BUT it must be structured to guarantee the same performance of this (and it doesn't)

```go
func publisherRoutine() {
    for {
        buf := make([]byte, 2048)
        readFrame(buf)

        for _, r := range reader {
           r.writeFrame(buf)
        }
    }
}
```
Otherwise users are gonna complain that "the old version was faster".

**Comment by Lakedaemon** [comment]
I see (please forgive me, I'm still trying though :( ). 

isn't "r.writeQueue <- buf" a blocking call, which might be what kills performance ? 
(me I don't know, but I'm googling. Better avoid locks, synchronization and blocking calls)

Coroutine are supposed to be really fast and lights (millions of them...so 50 coroutines should be really quick)

in there https://gobyexample.com/non-blocking-channel-operations
they suggest to use non blocking channel operations
(but his is go stuff, I don't understand ^^;)

Also, does r.writeQueue <- buf do a copy of the buffer ?
If yes, maybee it could be possible to have a single queue of reference/pointers to the buffers
and to only send to the readers the index of the buffer in this queue that should be sent

There is no reason there should be more garbage created with the coroutine versions thant with the simple thread and for loop version (the buffer is garbaged collected once it isn't used anymore in both coroutine/simple loop cases)

If the performance issue was resolved for the coroutine version, the server would scale with more cores and NOT suffer from slow readers

Best regards,
Olivier

**Comment by aler9** [comment]
> in there https://gobyexample.com/non-blocking-channel-operations
> they suggest to use non blocking channel operations
> (but his is go stuff, I don't understand ^^;)

yes, and this is the way to go, either this or ring buffers - see here
https://www.adaltas.com/en/2017/11/21/scaling-massive-real-time-data-pipelines-with-go/

> Also, does r.writeQueue <- buf do a copy of the buffer ?

no, it doesn't, byte buffers are always passed by reference in Go

**Comment by Lakedaemon** [comment]
Yeah, your link is much much better than the simple stuff I googled in just 1 minute.

Also, Indeed, I don't even need performance. my vps has 250mb/s and I'm just using 12mbits/s
sending only 100Ko/s for 50 people with 1 virtual Core with 2Go Ram
 
Reader blocking operations is a no go for me though (some students have fiber, others are on a lame connexion that they share with their family during lock downs)

Would it be better if the rtsp server was using tcp instead of udp ?

**Comment by dorinclisu** [comment]
> Otherwise users are gonna complain that "the old version was faster".

Perhaps the serial vs parallel behavior can be toggled from config file? Then it's up to the user if they want to give you performance or risk having the issue with one slow client.

**Comment by Lakedaemon** [comment]
that would be nice to have. Maybe, it would solve my issue (as I do not need performance that much)

**Comment by aler9** [comment]
Now readings and writings are performed in two separate routines, and laggy clients don't influence others anymore. This new version is about 5% slower, i think it's a fair price to pay to solve this issue.

**Comment by aler9** [comment]
added in v0.13.2

**Comment by Lakedaemon** [comment]
Many thanks, If laggy clients don't influence others anymore, I'll be able to use it in production. I'll test it this next week. 5% less performance is no problem as I could still accept 200% slower with my work load

Many many thanks. :)

**Comment by bbnp91** [comment]
Thanks for the update. 

Could you please help by advising how to effectively set the ReadBuffer. 
In previous versions 0.13.1 I was able to have around 500 subs of a 64kb audio stream via UDP.

Using the default config with a ReadBuffer of 512 in this version, I get a blocky decode. Increasing the ReadBuffer resolves the issue as expected.

Is there a general rule, or could you provide some insight as to how to best set the ReadBuffer for specific use cases, or at least one where the server is serving hundreds or thousands of clients from one publisher?

Thanks

**Comment by aler9** [comment]
I tuned readBufferCount to serve a single stream of 90 mbps with UDP. My advice to tune it is

* connect the maximum number of clients to the server (write a script that does it)
* increase readBufferCount until the RAM drains up

The only limit is RAM - increasing it is not a problem in terms of CPU, it's only a problem in terms of RAM.
